### PR TITLE
Error with better explanation when binary operation fails due to non-boolean masks

### LINKED ISSUE #2944: '==' operator between DataArray objects error when a mask with non-boolean values or unit given
When I tried to use '==' operator between DataArray object, it rasied an error in certain circumstances.

Here is the example code below. 

Given these context below,
```
import scipp as sc
import numpy as np

data = sc.array(dims=['x', 'y'], values = np.random.rand(4, 3))
```

This code works
```
mask0 = sc.array(dims=['x'], values=[False, True, True, False])
dummy_data_array = sc.DataArray(data, masks={'mask0': mask0})
dummy_data_array == dummy_data_array 
```

But these two below don't.
```
# When a mask has unit
mask0 = sc.array(dims=['x'], values=[False, True, True, False], unit='')
dummy_data_array = sc.DataArray(data, masks={'mask0': mask0})
dummy_data_array == dummy_data_array
```
```
# When a mask is non-boolean
mask0 = sc.array(dims=['x'], values=[1,2,3,4])
dummy_data_array = sc.DataArray(data, masks={'mask0': mask0})
dummy_data_array == dummy_data_array

```

These two cases raise an error like below.
```
---------------------------------------------------------------------------
UnitError                                 Traceback (most recent call last)
Cell In[24], line 3
      1 mask0 = sc.array(dims=['x'], values=[1,2,3,4])
      2 dummy_data_array = sc.DataArray(data, masks={'mask0': mask0})
----> 3 dummy_data_array == dummy_data_array

UnitError: Expected unit None, got dimensionless.
```

### PR DESCRIPTION
**Pull request by YooSunYoung**
Fixes #2944

### CODE REVIEW

**Review by jl-wynen — APPROVED** [review]

File: lib/dataset/dataset.cpp [hunk 1]
```diff
@@ -277,10 +277,16 @@ typename Masks::holder_type union_or(const Masks &currentMasks,
   for (const auto &[key, item] : otherMasks) {
     if (!currentMasks.contains(key))
       out.insert_or_assign(key, copy(item));
-    else if (out[key].dims().includes(item.dims()))
+    else if (item.dtype() != core::dtype<bool>) {
+      std::string optional_message = " This operation is not supported for "
+                                     "non-boolean masks with same names.";
```
Comment (SimonHeybrock) [inline lib/dataset/dataset.cpp:282]: Might be more useful to instead include the name of the offending mask, something like:

```
"Cannot combine non-boolean mask '" + key + "' in operation."
```

File: lib/dataset/dataset.cpp [hunk 2]
```diff
@@ -277,10 +277,16 @@ typename Masks::holder_type union_or(const Masks &currentMasks,
...
+      except::throw_mismatch_error(core::dtype<bool>, item.dtype(),
+                                   optional_message);
```
Comment (SimonHeybrock) [inline lib/dataset/dataset.cpp:284]: I think this is not catching the case where `otherMasks` is `bool` but `currentMasks` is not?

Comment (YooSunYoung) [reply lib/dataset/dataset.cpp:284]: You're right, it also doesn't catch if the non-boolean mask in the `otherMasks` is not in the `currentMasks`
I thought it only matters when you want to combine 2 non-boolean masks.

For example, if `data_array_1` has a non-boolean mask called `mask0` and `data_array_2` has a non-boolean mask called `mask1`, then they will be in the result of the operation will just have two non-boolean masks, `mask0` and `mask1`. 

Should we just forbid any non-boolean masks for the binary operations...?

Actually, after you mentioned that, I think it'll be weird to have non-boolean masks when you do `==` operation... Maybe we should just forbid it...?

File: lib/dataset/dataset.cpp (hunk 1 above)
Comment (YooSunYoung) [reply lib/dataset/dataset.cpp:282]: Yes...! I will do.

Comment (YooSunYoung) [reply lib/dataset/dataset.cpp:282]: I created a new PR in https://github.com/scipp/scipp/pull/2975

**Pull request merged**